# Final Examination

    Answer the question or complete the task outlined below, use the specific method described if applicable

    Note: Use [Alt + Enter] to execute the code.
    
    Deadline: January 20, 2021 11:59 PM (renatomaaliw3@yahoo.com)
    
    See yah guys next semester! Love yah!..

### 1. Import Numpy and Pandas

In [1]:
import numpy as np

In [2]:
import pandas as pd

### 2. What is

a. The modulo of 25 divided by 7

b. THe 11th root of 528

c. The natural logarithm of 58

d. $e^{12}$

In [1]:
# a. modulo

In [3]:
25%7

4

In [2]:
# b. 11th root

In [5]:
528**(1/11)

1.7681217836060341

In [3]:
# c. Natural logarithm

In [11]:
np.log(58)

4.060443010546419

In [4]:
# d. Euler's Number

In [15]:
np.exp(12)

162754.79141900392

### 3. Split the the string into a list

str = "Machine Learning next semester is a fun subject"

In [4]:
str = "Machine Learning next semester is a fun subject"

In [5]:
str.split()

['Machine', 'Learning', 'next', 'semester', 'is', 'a', 'fun', 'subject']

In [6]:
str_split=str.split()
str_split

['Machine', 'Learning', 'next', 'semester', 'is', 'a', 'fun', 'subject']

### 4. Extract the following strings from number 3

In [5]:
# Machine

In [31]:
str[:7]

'Machine'

In [32]:
str_split[0]

'Machine'

In [6]:
# Learning

In [33]:
str[8:16]

'Learning'

In [34]:
str_split[1]

'Learning'

In [7]:
# next

In [38]:
str[17:21]

'next'

In [39]:
str_split[2]

'next'

In [8]:
# semester

In [43]:
str[22:30]

'semester'

In [44]:
str_split[3]

'semester'

In [9]:
# subject

In [11]:
str[-7:]

'subject'

In [12]:
str_split[-1]

'subject'

### 5. Given the variables

name = "James"

work = "Programmer"

__Use .format to print the following string:__

James is a Programmer

In [15]:
name = "James"
work = "Programmer"

In [17]:
print('{} is a {}'.format(name, work))

James is a Programmer


### 6. Given the list below, use indexing to grab instances of 'target'

In [18]:
list_1 = [1,2,'target 1',[4,[5,6,7,[8,9,'target 2',[11,12,13,14,[15,'target 3']]],16,17]], 'target 4']

In [11]:
# target 1


In [19]:
list_1[2]

'target 1'

In [12]:
# target 2


In [20]:
list_1[3][1][3][2]

'target 2'

In [13]:
# target 3


In [23]:
list_1[3][1][3][3][4][1]

'target 3'

In [ ]:
# target 4


In [24]:
list_1[4]

'target 4'

### 7. Given the dictionary below, use indexing to grab instances of 'tricky'

In [26]:
d = {'k1':[1,'tricky 1',3,{'k2':['tricky 2', 4, 5,{'k3':[1,2,3,'tricky 3']}]}]}

In [27]:
# tricky 1


In [29]:
d['k1'][1]

'tricky 1'

In [17]:
# tricky 2


In [34]:
d['k1'][3]['k2'][0]

'tricky 2'

In [35]:
# tricky 3


In [40]:
d['k1'][3]['k2'][3]['k3'][3]

'tricky 3'

### 8. Create a function that grabs the mobile phones operator identity.

  0917-3453-234

e.g. So for example, passing '0917-3453-234' would return '0917'

In [41]:
# 0917


In [43]:
def phone_id(str):
    return str[:4]

phone_id('0917-3453-234')

'0917'

### 9. Create a function that counts the number of times a "cat" occurs in a string. Ignore edge cases.

In [50]:
str = "Felix is a cat. The cat is owned by Felice bought at the Cat Store"

In [60]:
def cat(str):
    
    list_split = str.lower().split()
    
    cat_count=0
    for cat_word in list_split:
        
        if cat_word=='cat': cat_count+=1
            
    return cat_count

cat(str)

2

### 10. Use lambda expressions and the filter() function to filter out words from a list that start with the letter 's' For example.

seq = ['silver', 'boat', 'shout', 'selenium', 'service', 'board']

In [61]:
# ['silver', 'shout', 'selenium', 'service']


In [62]:
seq = ['silver', 'boat', 'shout', 'selenium', 'service', 'board']


In [68]:
 list(filter(lambda x: x[0]=='s', seq))

['silver', 'shout', 'selenium', 'service']

### 11. Create the arrays using Numpy .arange()

In [28]:
# .arange


array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20])

In [72]:
np.arange(21)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20])

In [31]:
# .arange


array([5, 7, 9])

In [74]:
np.arange(5,10,2)

array([5, 7, 9])

In [32]:
# .arange


array([10, 13, 16, 19, 22, 25, 28, 31, 34, 37, 40, 43, 46, 49, 52, 55, 58,
       61, 64, 67, 70, 73, 76, 79, 82, 85, 88, 91, 94, 97])

In [78]:
np.arange(10,98,3)

array([10, 13, 16, 19, 22, 25, 28, 31, 34, 37, 40, 43, 46, 49, 52, 55, 58,
       61, 64, 67, 70, 73, 76, 79, 82, 85, 88, 91, 94, 97])

### 12. Create the array using np.ones

array([[5., 5., 5., 5., 5.],
       [5., 5., 5., 5., 5.],
       [5., 5., 5., 5., 5.],
       [5., 5., 5., 5., 5.],
       [5., 5., 5., 5., 5.]])

In [86]:
(np.ones(25).reshape(5,5))*5

array([[5., 5., 5., 5., 5.],
       [5., 5., 5., 5., 5.],
       [5., 5., 5., 5., 5.],
       [5., 5., 5., 5., 5.],
       [5., 5., 5., 5., 5.]])

### 13. Using .randn, create a 5 x 5 normally distributed array

In [87]:
np.random.seed(10) # execute this so that we can have the same values

In [89]:
np.random.randn(5,5)

array([[ 1.3315865 ,  0.71527897, -1.54540029, -0.00838385,  0.62133597],
       [-0.72008556,  0.26551159,  0.10854853,  0.00429143, -0.17460021],
       [ 0.43302619,  1.20303737, -0.96506567,  1.02827408,  0.22863013],
       [ 0.44513761, -1.13660221,  0.13513688,  1.484537  , -1.07980489],
       [-1.97772828, -1.7433723 ,  0.26607016,  2.38496733,  1.12369125]])

array([[ 1.3315865 ,  0.71527897, -1.54540029, -0.00838385,  0.62133597],
       [-0.72008556,  0.26551159,  0.10854853,  0.00429143, -0.17460021],
       [ 0.43302619,  1.20303737, -0.96506567,  1.02827408,  0.22863013],
       [ 0.44513761, -1.13660221,  0.13513688,  1.484537  , -1.07980489],
       [-1.97772828, -1.7433723 ,  0.26607016,  2.38496733,  1.12369125]])

### 14. Given the array below, identify the maximum and minimum values

In [90]:
np.random.seed(10)
random_array = np.random.randint(0,25,10)
random_array

array([ 9,  4, 15,  0, 17, 16, 17,  8,  9,  0])

In [91]:
# maximum value?


In [92]:
np.max(random_array)

17

In [93]:
# minimum value?


In [94]:
np.min(random_array)

0

### 15. Given the array below, slice and select the following

In [95]:
new_arr = np.arange(0,16)
new_arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])

array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])

In [96]:
new_arr[5:]

array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])

array([2, 3, 4])

In [97]:
new_arr[2:5]

array([2, 3, 4])

array([11, 12])

In [98]:
new_arr[11:13]

array([11, 12])

### 16. Using NumPy create the following Matrices

$ a = \begin{bmatrix} 1 & -3 \\ 5 & 6 \end{bmatrix}$

$ b = \begin{bmatrix} 6 & 7 & -2 \\ 4 & 5 & 8 \end{bmatrix}$

In [75]:
# matrix a

In [104]:
a = np.array([1,-3,5,6]).reshape(2,2)
a

array([[ 1, -3],
       [ 5,  6]])

In [105]:
# matrix b

In [107]:
b = np.array([[6,7,-2],[4,5,6]])
b

array([[ 6,  7, -2],
       [ 4,  5,  6]])

### 17. Create the following NumPy array and then select the following submatrices:

$ c = \begin{bmatrix} 2 & 4 & 6 & 8 & 10 & 12 & 14 & 16 \\ 18 & 20 & 22 & 24 & 26 & 28 & 30 & 32 \\ 34 & 36 & 38 & 40 & 42 & 44 & 46 & 48 \\ 50 & 52 & 54 & 56 & 58 & 60 & 62 & 64\end{bmatrix}$

In [3]:
c=np.arange(2,65,2).reshape(4,8)
c

array([[ 2,  4,  6,  8, 10, 12, 14, 16],
       [18, 20, 22, 24, 26, 28, 30, 32],
       [34, 36, 38, 40, 42, 44, 46, 48],
       [50, 52, 54, 56, 58, 60, 62, 64]])

$ subd1 = \begin{bmatrix} 24 & 26 & 28 & 30 & 32 \\ 40 & 42 & 44 & 46 & 48\end{bmatrix}$

In [4]:
subd1 = c[1:3,3:]
subd1

array([[24, 26, 28, 30, 32],
       [40, 42, 44, 46, 48]])

$ subd2 = \begin{bmatrix} 22 & 24 & 26 & 30 \\ 38 & 40 & 42 & 46\end{bmatrix}$

In [63]:
subd2 = np.transpose(np.concatenate([c[1:3, 2:3],c[1:3, 3:4],c[1:3, 4:5], c[1:3, 6:7]]).reshape(4,2))
subd2

array([[22, 24, 26, 30],
       [38, 40, 42, 46]])

$ subd3 = \begin{bmatrix} 26 \\ 42 \\ 58 \end{bmatrix}$

In [52]:
subd3 = c[1:5,4:5]
subd3

array([[26],
       [42],
       [58]])

$ subd5 = \begin{bmatrix} 18 & 34 \\ 50 & 14 \\ 30 & 46 \end{bmatrix}$ hint: use numpy.concatenate and numpy.transpose

In [67]:
subd5 = np.concatenate([c[1:,:1],c[:3,-2:-1]]).reshape(3,2)
subd5

array([[18, 34],
       [50, 14],
       [30, 46]])

### 18. Matrix Operations

$ e = \begin{bmatrix} 3 & -3 \\ 1 & -2 \\ 6 & 4 \end{bmatrix}$ 

$ f = \begin{bmatrix} 1 & 5 \\ -3 & 5 \\ 2 & 5 \end{bmatrix}$

In [69]:
e=np.array([[3,-3],[1,-2],[6,4]])
e

array([[ 3, -3],
       [ 1, -2],
       [ 6,  4]])

In [71]:
f=np.array([1,5,-3,5,2,5]).reshape(3,2)
f

array([[ 1,  5],
       [-3,  5],
       [ 2,  5]])

In [72]:
# add matrix e and matrix f

In [74]:
np.add(e,f)

array([[ 4,  2],
       [-2,  3],
       [ 8,  9]])

In [79]:
# multipy matrix e by 5 (scalar multiplication)

In [75]:
np.multiply(e,5)

array([[ 15, -15],
       [  5, -10],
       [ 30,  20]])

In [80]:
# multipy matrix e by matrix f

In [77]:
np.dot(e,f)

ValueError: shapes (3,2) and (3,2) not aligned: 2 (dim 1) != 3 (dim 0)

In [79]:
#to avoid error, transpose first 1 matrix, in this case I transposed f po
np.dot(e,f.transpose())

array([[-12, -24,  -9],
       [ -9, -13,  -8],
       [ 26,   2,  32]])

In [81]:
# transpose matrix f

In [78]:
np.transpose(f)

array([[ 1, -3,  2],
       [ 5,  5,  5]])

### 19. Find the determinant of the Matrix

 $ h = \begin{bmatrix} 2 & -3 & 4 \\ 2 & -3 & 2 \\ -1 & -4 & 3 \end{bmatrix} $

In [80]:
h=np.array([2,-3,4,2,-3,2,-1,-4,3]).reshape(3,3)
h

array([[ 2, -3,  4],
       [ 2, -3,  2],
       [-1, -4,  3]])

In [81]:
np.linalg.det(h)

-22.000000000000004

In [82]:
np.linalg.det(h).astype('int32')

-22

### 20. Find the Inverse of the Matrix and prove that it is the Inverse of the Original

 $ i = \begin{bmatrix} 4 & 1 & 5 \\ 0 & 3 & -3 \\ -2 & 2 & -2 \end{bmatrix} $

In [83]:
i=np.array([4,1,5,0,3,-3,-2,2,-2]).reshape(3,3)
i

array([[ 4,  1,  5],
       [ 0,  3, -3],
       [-2,  2, -2]])

In [85]:
i_inv = np.linalg.inv(i)
i_inv

array([[ 1.38777878e-17,  3.33333333e-01, -5.00000000e-01],
       [ 1.66666667e-01,  5.55555556e-02,  3.33333333e-01],
       [ 1.66666667e-01, -2.77777778e-01,  3.33333333e-01]])

In [86]:
#to prove that, the following is computed...
np.dot(i,i_inv).astype('int32')

array([[1, 0, 0],
       [0, 1, 0],
       [0, 0, 0]])

In [87]:
# i * i_inv = identity matrix , therefore it is the inverse of the original

### 21. Solve for the value of x:  $ e^{4x} = 132 $

In [88]:
np.log(132)/4

1.2207004806465926

### 22. Solve the following problem using Systems of Linear Equations

1. The ticket for a ride is 1.50 (USD) for children and 4.00 (USD) for adults. During some instances, 2200 customers have enter the carnival and 5050 (USD) is collected. How many adults and children went to pay for the carnival?

### Let x be children, y be adults

$ 1.5x + 4y = 5050 $ ;ticket prices

$ x + y = 2200 $ ;customers

In [89]:
a=np.array([[1.5,4],[1,1]])
a

array([[1.5, 4. ],
       [1. , 1. ]])

In [90]:
b=np.array([[5050],[2200]])
b

array([[5050],
       [2200]])

In [91]:
np.linalg.solve(a,b)

array([[1500.],
       [ 700.]])

In [92]:
#therefore, there are 1500 children and 700 adults who went to pay for the carnival

2. At a store, May pays 34 (USD) for 2 pounds of apples, 1 pound of berries, and 4 pounds of cherries. Tom pays 35 (USD) for 3 pounds of apples, 2 pounds of berries, and 2 pounds of cherries. Lee pays 49 (USD) for 5 pounds of apples, 3 pounds of berries, and 2 pounds of cherries. What is the price per pound for apples, for berries, and for cheeries?


### Let x be apples, y be berries, z be cherries

$ 2x + y + 4z = 34 $ ;May

$ 3x + 2y + 2z = 35 $ ;Tom

$ 5x + 3y + 2z = 49 $ ;Lee


In [93]:
a = np.array([2,1,4,3,2,2,5,3,2]).reshape(3,3)
a

array([[2, 1, 4],
       [3, 2, 2],
       [5, 3, 2]])

In [94]:
b=np.array([[34],[35],[49]])
b

array([[34],
       [35],
       [49]])

In [95]:
np.linalg.solve(a,b)

array([[3.],
       [8.],
       [5.]])

In [96]:
#therefore the price per pound for apples is 3 USD, berries is 8 USD, and cherries is 5 USD

### 23. Find the Eigen Values and Eigen Vectors of the Matrix Below

$ eg = \begin{bmatrix} 3 & -2 \\ -5 & 3 \end{bmatrix} $

In [98]:
eg=np.array([3,-2,-5,3]).reshape(2,2)
eg

array([[ 3, -2],
       [-5,  3]])

In [99]:
np.linalg.eig(eg)

(array([ 6.16227766, -0.16227766]),
 array([[ 0.53452248,  0.53452248],
        [-0.84515425,  0.84515425]]))

### To find unnormalized vector:
### for eigen value of 6.16

$ B = \begin{bmatrix} 3-6.16227766 & -2 \\ -5 & 3-6.16227766 \end{bmatrix} = \begin{bmatrix} -3.16227766 & -2 \\ -5 & -3.16227766 \end{bmatrix} $

### Let x = 1

$ -3.16227766(1) - 2y = 0 ; y=-1.58113883 $

$ -5(1) -3.16227766y = 0 ; y=-1.5811388301683793 $

In [103]:
#thus, for eigen value of 6.16, unnormalized vector is...
un_eig6=np.array([[1],[-1.58113883]])
un_eig6

array([[ 1.        ],
       [-1.58113883]])

In [105]:
#to prove that, normalized vector is computed by the following...
euclid_val = np.sqrt(np.sum(np.square(un_eig6)))
euclid_val

1.8708286933158174

In [106]:
norm_vec1 = un_eig6/euclid_val
norm_vec1

array([[ 0.53452248],
       [-0.84515425]])

In [107]:
np.linalg.eig(eg)

(array([ 6.16227766, -0.16227766]),
 array([[ 0.53452248,  0.53452248],
        [-0.84515425,  0.84515425]]))

In [108]:
#norm_vec1 is same as computed by python

### To find unnormalized vector:
### for eigen value of -0.16

$ B = \begin{bmatrix} 3-(-0.16227766) & -2 \\ -5 & 3-(-0.16227766) \end{bmatrix} = \begin{bmatrix} 3.16227766 & -2 \\ -5 & 3.16227766 \end{bmatrix} $

### Let x = 1

$ 3.16227766(1) - 2y = 0 ; y= 1.58113883 $

$ -5(1) + 3.16227766y = 0 ; y= 1.5811388301683793 $

In [111]:
#thus, for eigen value of -0.16, unnormalized vector is...
un_eig0=np.array([[1],[1.58113883]])
un_eig0

array([[1.        ],
       [1.58113883]])

In [113]:
#to prove that, normalized vector is computed by the following...
euclid_val2 = np.sqrt(np.sum(np.square(un_eig0)))
euclid_val2

1.8708286933158174

In [114]:
norm_vec2 = un_eig0/euclid_val2
norm_vec2

array([[0.53452248],
       [0.84515425]])

In [115]:
np.linalg.eig(eg)

(array([ 6.16227766, -0.16227766]),
 array([[ 0.53452248,  0.53452248],
        [-0.84515425,  0.84515425]]))

In [116]:
#norm_vec2 is same as computed by python

### 24. Create a Series object out of the following:

a. numpy array with [2,4,6,8,12]

b. dictionary {'Ilocos': 38, 'Bicol': 26, 'Quezon': 19, 'Batangas': 55}

In [118]:
# from numpy array


In [122]:
np_array = np.array([2,4,6,8,12])
np_array

array([ 2,  4,  6,  8, 12])

In [123]:
pd_series = pd.Series(np_array)
pd_series

0     2
1     4
2     6
3     8
4    12
dtype: int32

0     2
1     4
2     6
3     8
4    12
dtype: int32

In [124]:
# from dictionary


In [125]:
dict = {'Ilocos': 38, 'Bicol': 26, 'Quezon': 19, 'Batangas': 55}
dict

{'Ilocos': 38, 'Bicol': 26, 'Quezon': 19, 'Batangas': 55}

In [126]:
pd_seriesd = pd.Series(dict)
pd_seriesd

Ilocos      38
Bicol       26
Quezon      19
Batangas    55
dtype: int64

Ilocos      38
Bicol       26
Quezon      19
Batangas    55
dtype: int64

### 25. Create the data frame as seen below

In [138]:
data_matrix = np.array([[423967, 38332521, 'CA'],[695662, 26448193, 'TX'],[141297, 19651127, 'NY'],[170312, 19552860, 'FL'],
             [149995, 12882135, 'IL']])
data_matrix

array([['423967', '38332521', 'CA'],
       ['695662', '26448193', 'TX'],
       ['141297', '19651127', 'NY'],
       ['170312', '19552860', 'FL'],
       ['149995', '12882135', 'IL']], dtype='<U11')

In [134]:
column_label = ['area', 'population', 'code']
row_label = ['California', 'Texas', 'New York', 'Florida', 'Illinois']


In [139]:
df = pd.DataFrame(data_matrix, row_label, column_label)
df

area population code
California  423967   38332521   CA
Texas       695662   26448193   TX
New York    141297   19651127   NY
Florida     170312   19552860   FL
Illinois    149995   12882135   IL

area  population code
California  423967    38332521   CA
Texas       695662    26448193   TX
New York    141297    19651127   NY
Florida     170312    19552860   FL
Illinois    149995    12882135   IL

### 26. Create another column 'density' based on number 25

In [140]:
df['density'] = [90.413926, 38.018740, 139.076746, 114.806121, 85.883763]
df

area population code     density
California  423967   38332521   CA   90.413926
Texas       695662   26448193   TX   38.018740
New York    141297   19651127   NY  139.076746
Florida     170312   19552860   FL  114.806121
Illinois    149995   12882135   IL   85.883763

area  population code     density
California  423967    38332521   CA   90.413926
Texas       695662    26448193   TX   38.018740
New York    141297    19651127   NY  139.076746
Florida     170312    19552860   FL  114.806121
Illinois    149995    12882135   IL   85.883763

### 26. Permanently drop the 'Texas' row based on number 25

In [141]:
df.drop('Texas', axis = 0, inplace = True)
df

area population code     density
California  423967   38332521   CA   90.413926
New York    141297   19651127   NY  139.076746
Florida     170312   19552860   FL  114.806121
Illinois    149995   12882135   IL   85.883763

area  population code     density
California  423967    38332521   CA   90.413926
New York    141297    19651127   NY  139.076746
Florida     170312    19552860   FL  114.806121
Illinois    149995    12882135   IL   85.883763

### 27. Create another Data Frame

In [226]:
np.random.seed(10) # execute this to have the same random values

In [221]:
np.random.randint(0,20,64)

array([ 9,  4, 15,  0, 17, 16, 17,  8,  9,  0, 10,  8,  4, 19, 16,  4, 15,
       11, 11,  1,  8,  4, 14, 17, 19, 13,  5, 13, 19, 13, 12,  1,  4, 18,
       13, 11, 10,  9, 15, 18, 16,  7, 11, 17, 14,  7, 11,  1,  0, 12,  5,
        4,  7, 18, 12,  2,  8, 15, 13, 15,  6, 10, 11, 16])

In [223]:
np.random.randint(0,20,64)

array([ 9,  4, 15,  0, 17, 16, 17,  8,  9,  0, 10,  8,  4, 19, 16,  4, 15,
       11, 11,  1,  8,  4, 14, 17, 19, 13,  5, 13, 19, 13, 12,  1,  4, 18,
       13, 11, 10,  9, 15, 18, 16,  7, 11, 17, 14,  7, 11,  1,  0, 12,  5,
        4,  7, 18, 12,  2,  8, 15, 13, 15,  6, 10, 11, 16])

In [230]:
#sir, di po makuha ung same randomly generated data kagaya sa baba, ibang data po nakukuha, proof po ung dalawang execution 
# na ginamit ung seed

In [227]:
df2 = pd.DataFrame(np.random.randint(0,20,64).reshape(8,8), ['R1','R2','R3','R4','R5','R6','R7','R8'], 
                   ['C1','C2','C3','C4','C5','C6','C7','C8'])
df2

C1  C2  C3  C4  C5  C6  C7  C8
R1   9   4  15   0  17  16  17   8
R2   9   0  10   8   4  19  16   4
R3  15  11  11   1   8   4  14  17
R4  19  13   5  13  19  13  12   1
R5   4  18  13  11  10   9  15  18
R6  16   7  11  17  14   7  11   1
R7   0  12   5   4   7  18  12   2
R8   8  15  13  15   6  10  11  16

In [228]:
#to copy the dataframe, manual na lang po pag create

In [229]:
np_array2 = np.array([8,3,7,15,16,10,2,2,2,14,2,17,16,15,4,11,16,9,2,12,4,1,13,19,4,4,3,7,17,15,1,14,7,16,2,9,19,2,14,17,16,15,
                      7,13,6,12,18,0,2,10,17,8,13,10,17,4,18,8,19,14,0,13,12,10]).reshape(8,8)
df2 = pd.DataFrame(np_array2, ['R1','R2','R3','R4','R5','R6','R7','R8'], 
                   ['C1','C2','C3','C4','C5','C6','C7','C8'])
df2

C1  C2  C3  C4  C5  C6  C7  C8
R1   8   3   7  15  16  10   2   2
R2   2  14   2  17  16  15   4  11
R3  16   9   2  12   4   1  13  19
R4   4   4   3   7  17  15   1  14
R5   7  16   2   9  19   2  14  17
R6  16  15   7  13   6  12  18   0
R7   2  10  17   8  13  10  17   4
R8  18   8  19  14   0  13  12  10

C1  C2  C3  C4  C5  C6  C7  C8
R1   8   3   7  15  16  10   2   2
R2   2  14   2  17  16  15   4  11
R3  16   9   2  12   4   1  13  19
R4   4   4   3   7  17  15   1  14
R5   7  16   2   9  19   2  14  17
R6  16  15   7  13   6  12  18   0
R7   2  10  17   8  13  10  17   4
R8  18   8  19  14   0  13  12  10

### 28. Select the following parts of the Data Frame from number 27 (use .loc or iloc)

In [233]:
df2.loc['R1':'R5', 'C2':'C6']

C2  C3  C4  C5  C6
R1   3   7  15  16  10
R2  14   2  17  16  15
R3   9   2  12   4   1
R4   4   3   7  17  15
R5  16   2   9  19   2

C2  C3  C4  C5  C6
R1   3   7  15  16  10
R2  14   2  17  16  15
R3   9   2  12   4   1
R4   4   3   7  17  15
R5  16   2   9  19   2

In [236]:
df2.iloc[2:8,2:7]

C3  C4  C5  C6  C7
R3   2  12   4   1  13
R4   3   7  17  15   1
R5   2   9  19   2  14
R6   7  13   6  12  18
R7  17   8  13  10  17
R8  19  14   0  13  12

C3  C4  C5  C6  C7
R3   2  12   4   1  13
R4   3   7  17  15   1
R5   2   9  19   2  14
R6   7  13   6  12  18
R7  17   8  13  10  17
R8  19  14   0  13  12

In [242]:
df2.iloc[:,0]

R1     8
R2     2
R3    16
R4     4
R5     7
R6    16
R7     2
R8    18
Name: C1, dtype: int32

R1     8
R2     2
R3    16
R4     4
R5     7
R6    16
R7     2
R8    18
Name: C1, dtype: int32

In [241]:
df2.iloc[2,:]

C1    16
C2     9
C3     2
C4    12
C5     4
C6     1
C7    13
C8    19
Name: R3, dtype: int32

C1    16
C2     9
C3     2
C4    12
C5     4
C6     1
C7    13
C8    19
Name: R3, dtype: int32

In [244]:
df2.iloc[5:7,3:7]

C4  C5  C6  C7
R6  13   6  12  18
R7   8  13  10  17

C4  C5  C6  C7
R6  13   6  12  18
R7   8  13  10  17

### 29. Fill in the Missing Data

In [245]:
# Create the Data Frame below

In [256]:
df3 = pd.DataFrame(np.array([[np.nan, 10, 25], [25, 20, np.nan],[85, 30, 55],[50, np.nan, np.nan],[np.nan, 50, np.nan]]),
                    [0, 1, 2, 3, 4],['A','B','C'])
df3['D'] = [20,40,60,80,100]
df3

A     B     C    D
0   NaN  10.0  25.0   20
1  25.0  20.0   NaN   40
2  85.0  30.0  55.0   60
3  50.0   NaN   NaN   80
4   NaN  50.0   NaN  100

A     B     C    D
0   NaN  10.0  25.0   20
1  25.0  20.0   NaN   40
2  85.0  30.0  55.0   60
3  50.0   NaN   NaN   80
4   NaN  50.0   NaN  100

In [255]:
# Replace missing values of column A using .mean

In [261]:
df3['A'] = df3['A'].fillna(value = df3['A'].mean())
df3

A     B     C    D
0  53.333333  10.0  25.0   20
1  25.000000  20.0   NaN   40
2  85.000000  30.0  55.0   60
3  50.000000   NaN   NaN   80
4  53.333333  50.0   NaN  100

In [262]:
# Replace missing values of column B using .median

In [263]:
df3['B'] = df3['B'].fillna(value = df3['B'].median())
df3

A     B     C    D
0  53.333333  10.0  25.0   20
1  25.000000  20.0   NaN   40
2  85.000000  30.0  55.0   60
3  50.000000  25.0   NaN   80
4  53.333333  50.0   NaN  100

In [264]:
# Replace missing values of column C using random numbers (whole numbers only) from 20 to 60, Hint: use the .apply() method

In [287]:
df3['C'] = df3.apply(lambda dfr: np.random.randint(20,61) if np.isnan(dfr['C']) else dfr['C'], axis = 1)
df3

A     B     C    D
0  53.333333  10.0  25.0   20
1  25.000000  20.0  60.0   40
2  85.000000  30.0  55.0   60
3  50.000000  25.0  53.0   80
4  53.333333  50.0  60.0  100

### 30. GroupBy

In [289]:
df4 = pd.read_csv('advertising.csv')
df4

Daily Time Spent on Site  Age  Area Income  Daily Internet Usage  \
0                       68.95   35     61833.90                256.09   
1                       80.23   31     68441.85                193.77   
2                       69.47   26     59785.94                236.50   
3                       74.15   29     54806.18                245.89   
4                       68.37   35     73889.99                225.58   
..                        ...  ...          ...                   ...   
995                     72.97   30     71384.57                208.58   
996                     51.30   45     67782.17                134.42   
997                     51.63   51     42415.72                120.37   
998                     55.55   19     41920.79                187.95   
999                     45.01   26     29875.80                178.35   

                             Ad Topic Line            City  Male  \
0       Cloned 5thgeneration orchestration     Wrightburgh     0   
1       Monitored national standardization       West Jodi     1   
2         Organic bottom-line service-desk        Davidton     0   
3    Triple-buffered reciprocal time-frame  West Terrifurt     1   
4            Robust logistical utilization    South Manuel     0   
..                                     ...             ...   ...   
995          Fundamental modular algorithm       Duffystad     1   
996        Grass-roots cohesive monitoring     New Darlene     1   
997           Expanded intangible solution   South Jessica     1   
998   Proactive bandwidth-monitored policy     West Steven     0   
999        Virtual 5thgeneration emulation     Ronniemouth     0   

                    Country        Timestamp  Clicked on Ad  
0                   Tunisia   3/27/2016 0:53              0  
1                     Nauru    4/4/2016 1:39              0  
2                San Marino  3/13/2016 20:35              0  
3                     Italy   1/10/2016 2:31              0  
4                   Iceland    6/3/2016 3:36              0  
..                      ...              ...            ...  
995                 Lebanon  2/11/2016 21:49              1  
996  Bosnia and Herzegovina   4/22/2016 2:07              1  
997                Mongolia   2/1/2016 17:24              1  
998               Guatemala   3/24/2016 2:35              0  
999                  Brazil   6/3/2016 21:43              1  

[1000 rows x 10 columns]

In [8]:
# Load data set advertising.csv (see our github Data Set folder)



Daily Time Spent on Site  Age  Area Income  Daily Internet Usage  \
0                       68.95   35     61833.90                256.09   
1                       80.23   31     68441.85                193.77   
2                       69.47   26     59785.94                236.50   
3                       74.15   29     54806.18                245.89   
4                       68.37   35     73889.99                225.58   
..                        ...  ...          ...                   ...   
995                     72.97   30     71384.57                208.58   
996                     51.30   45     67782.17                134.42   
997                     51.63   51     42415.72                120.37   
998                     55.55   19     41920.79                187.95   
999                     45.01   26     29875.80                178.35   

                             Ad Topic Line            City  Male  \
0       Cloned 5thgeneration orchestration     Wrightburgh     0   
1       Monitored national standardization       West Jodi     1   
2         Organic bottom-line service-desk        Davidton     0   
3    Triple-buffered reciprocal time-frame  West Terrifurt     1   
4            Robust logistical utilization    South Manuel     0   
..                                     ...             ...   ...   
995          Fundamental modular algorithm       Duffystad     1   
996        Grass-roots cohesive monitoring     New Darlene     1   
997           Expanded intangible solution   South Jessica     1   
998   Proactive bandwidth-monitored policy     West Steven     0   
999        Virtual 5thgeneration emulation     Ronniemouth     0   

                    Country        Timestamp  Clicked on Ad  
0                   Tunisia   3/27/2016 0:53              0  
1                     Nauru    4/4/2016 1:39              0  
2                San Marino  3/13/2016 20:35              0  
3                     Italy   1/10/2016 2:31              0  
4                   Iceland    6/3/2016 3:36              0  
..                      ...              ...            ...  
995                 Lebanon  2/11/2016 21:49              1  
996  Bosnia and Herzegovina   4/22/2016 2:07              1  
997                Mongolia   2/1/2016 17:24              1  
998               Guatemala   3/24/2016 2:35              0  
999                  Brazil   6/3/2016 21:43              1  

[1000 rows x 10 columns]

### 31. Based on the loaded data set, show the following:

    - use groupby, sort_values, head

In [291]:
df4.groupby('Male').mean()['Age']

Male
0    36.186898
1    35.817048
Name: Age, dtype: float64

In [10]:
# Average age of Male (1) and Female (0)



Male
0    36.186898
1    35.817048
Name: Age, dtype: float64

In [472]:
df4.groupby('Country').max().sort_values('Daily Time Spent on Site', ascending = False).head(1)['Daily Time Spent on Site']

Country
French Polynesia    91.43
Name: Daily Time Spent on Site, dtype: float64

In [26]:
# Country with the highest daily time spent on the site



Country
French Polynesia    91.43
Name: Daily Time Spent on Site, dtype: float64

In [632]:
df4[['Country','Daily Time Spent on Site']].groupby('Country').max().sort_values('Daily Time Spent on Site').head(10)['Daily Time Spent on Site']

Country
Kiribati                                               36.37
Marshall Islands                                       43.16
Sao Tome and Principe                                  44.49
Lithuania                                              46.98
Romania                                                49.99
Saint Kitts and Nevis                                  50.52
Andorra                                                53.33
British Indian Ocean Territory (Chagos Archipelago)    54.70
New Caledonia                                          56.30
Montserrat                                             56.64
Name: Daily Time Spent on Site, dtype: float64

In [25]:
# Top 10 countries with the lowest daily time spent on the site


Country
Kiribati                                               36.37
Marshall Islands                                       43.16
Sao Tome and Principe                                  44.49
Lithuania                                              46.98
Romania                                                49.99
Saint Kitts and Nevis                                  50.52
Andorra                                                53.33
British Indian Ocean Territory (Chagos Archipelago)    54.70
New Caledonia                                          56.30
Montserrat                                             56.64
Name: Daily Time Spent on Site, dtype: float64

In [442]:
df4.groupby('Male').count()['Age']

Male
0    519
1    481
Name: Age, dtype: int64

In [20]:
# How many Male (1) and Female are on the data set



Male
0    519
1    481
Name: Age, dtype: int64

In [445]:
df4.groupby('City').sum().sort_values('Daily Internet Usage', ascending=False).head(5)['Daily Internet Usage']

City
Williamsport       520.02
Lisamouth          491.30
Wrightburgh        449.89
Benjaminchester    434.68
Lake Patrick       430.09
Name: Daily Internet Usage, dtype: float64

In [37]:
# Which top 5 city in total has the highest daily internet usage



City
Williamsport       520.02
Lisamouth          491.30
Wrightburgh        449.89
Benjaminchester    434.68
Lake Patrick       430.09
Name: Daily Internet Usage, dtype: float64

In [634]:
df4.groupby('Country').sum().sort_values('Clicked on Ad', ascending = False).head(3)['Clicked on Ad']

Country
Australia    7
Turkey       7
Ethiopia     7
Name: Clicked on Ad, dtype: int64

In [42]:
# Top 3 Country that clicked on an ad



Country
Australia    7
Turkey       7
Ethiopia     7
Name: Clicked on Ad, dtype: int64

### 32. Exploratory Data Analysis

    - you can use any method as long as the answer is correct
    - you can use combination of functions, groupby, data frame filtering, etch., whatever method you will prefer
    - remember if you are confused do not force to code at once, break down the problem into pieces
    - i will not show you the answers, you are on your own on this one

In [494]:
df5 = pd.read_csv('summer.csv')
df5

Year    City      Sport           Discipline  \
0      1896  Athens   Aquatics             Swimming   
1      1896  Athens   Aquatics             Swimming   
2      1896  Athens   Aquatics             Swimming   
3      1896  Athens   Aquatics             Swimming   
4      1896  Athens   Aquatics             Swimming   
...     ...     ...        ...                  ...   
31160  2012  London  Wrestling  Wrestling Freestyle   
31161  2012  London  Wrestling  Wrestling Freestyle   
31162  2012  London  Wrestling  Wrestling Freestyle   
31163  2012  London  Wrestling  Wrestling Freestyle   
31164  2012  London  Wrestling  Wrestling Freestyle   

                         Athlete Country Gender                       Event  \
0                  HAJOS, Alfred     HUN    Men              100M Freestyle   
1               HERSCHMANN, Otto     AUT    Men              100M Freestyle   
2              DRIVAS, Dimitrios     GRE    Men  100M Freestyle For Sailors   
3             MALOKINIS, Ioannis     GRE    Men  100M Freestyle For Sailors   
4             CHASAPIS, Spiridon     GRE    Men  100M Freestyle For Sailors   
...                          ...     ...    ...                         ...   
31160         JANIKOWSKI, Damian     POL    Men                    Wg 84 KG   
31161  REZAEI, Ghasem Gholamreza     IRI    Men                    Wg 96 KG   
31162             TOTROV, Rustam     RUS    Men                    Wg 96 KG   
31163          ALEKSANYAN, Artur     ARM    Men                    Wg 96 KG   
31164             LIDBERG, Jimmy     SWE    Men                    Wg 96 KG   

        Medal  
0        Gold  
1      Silver  
2      Bronze  
3        Gold  
4      Silver  
...       ...  
31160  Bronze  
31161    Gold  
31162  Silver  
31163  Bronze  
31164  Bronze  

[31165 rows x 9 columns]

In [46]:
# Load data set summer.csv (see our github Data Set folder)


Year    City      Sport           Discipline  \
0      1896  Athens   Aquatics             Swimming   
1      1896  Athens   Aquatics             Swimming   
2      1896  Athens   Aquatics             Swimming   
3      1896  Athens   Aquatics             Swimming   
4      1896  Athens   Aquatics             Swimming   
...     ...     ...        ...                  ...   
31160  2012  London  Wrestling  Wrestling Freestyle   
31161  2012  London  Wrestling  Wrestling Freestyle   
31162  2012  London  Wrestling  Wrestling Freestyle   
31163  2012  London  Wrestling  Wrestling Freestyle   
31164  2012  London  Wrestling  Wrestling Freestyle   

                         Athlete Country Gender                       Event  \
0                  HAJOS, Alfred     HUN    Men              100M Freestyle   
1               HERSCHMANN, Otto     AUT    Men              100M Freestyle   
2              DRIVAS, Dimitrios     GRE    Men  100M Freestyle For Sailors   
3             MALOKINIS, Ioannis     GRE    Men  100M Freestyle For Sailors   
4             CHASAPIS, Spiridon     GRE    Men  100M Freestyle For Sailors   
...                          ...     ...    ...                         ...   
31160         JANIKOWSKI, Damian     POL    Men                    Wg 84 KG   
31161  REZAEI, Ghasem Gholamreza     IRI    Men                    Wg 96 KG   
31162             TOTROV, Rustam     RUS    Men                    Wg 96 KG   
31163          ALEKSANYAN, Artur     ARM    Men                    Wg 96 KG   
31164             LIDBERG, Jimmy     SWE    Men                    Wg 96 KG   

        Medal  
0        Gold  
1      Silver  
2      Bronze  
3        Gold  
4      Silver  
...       ...  
31160  Bronze  
31161    Gold  
31162  Silver  
31163  Bronze  
31164  Bronze  

[31165 rows x 9 columns]

In [495]:
# How many cities host the Olympics based from the data set?



In [497]:
df5['City'].nunique()

22

In [ ]:
#22 cities

In [498]:
# What are the sports competed based from the data set?



In [781]:
df5.groupby('Sport').count()[[]]

Empty DataFrame
Columns: []
Index: [Aquatics, Archery, Athletics, Badminton, Baseball, Basketball, Basque Pelota, Boxing, Canoe, Canoe / Kayak, Cricket, Croquet, Cycling, Equestrian, Fencing, Football, Golf, Gymnastics, Handball, Hockey, Ice Hockey, Jeu de paume, Judo, Lacrosse, Modern Pentathlon, Polo, Rackets, Roque, Rowing, Rugby, Sailing, Shooting, Skating, Softball, Table Tennis, Taekwondo, Tennis, Triathlon, Tug of War, Volleyball, Water Motorsports, Weightlifting, Wrestling]

In [516]:
# What is the name of the athlete who hauls the most gold medal for the country USA?



In [763]:
df5[(df5['Country'] == 'USA') &  (df5['Medal'] == 'Gold')].groupby(by='Athlete').count().sort_values(by='Medal', ascending=False).head(1)[[]]

Empty DataFrame
Columns: []
Index: [PHELPS, Michael]

In [589]:
# What are the top 5 countries who hauls the most medal (Gold, Silver, Bronze) in Weightlifting?



In [758]:
df5[df5['Sport']=='Weightlifting'].groupby('Country').count().sort_values('Medal', ascending=False).head(5)[[]]

Empty DataFrame
Columns: []
Index: [URS, CHN, USA, BUL, POL]

In [599]:
# What is the name of the top 3 athletes who hauls the most gold medal for the country CHN in Gymnastics?



In [761]:
df5[(df5['Country'] == 'CHN') &  (df5['Medal'] == 'Gold')].groupby(by='Athlete').count().sort_values(by='Medal', ascending=False).head(3)[[]]

Empty DataFrame
Columns: []
Index: [ZOU, Kai, FU, Mingxia, GUO, Jingjing]

In [671]:
# Which country has the 3rd highest collective Bronze medal tally in the dataset?



In [770]:
df5[(df5['Medal'] == 'Bronze')].groupby(by='Country').count().sort_values(by='Medal', ascending=False).head(3).iloc[2:][[]]

Empty DataFrame
Columns: []
Index: [GBR]

In [696]:
# How many Men and Women participated in the Volleyball event during the Los Angeles Olympics?



In [789]:
df5[(df5['Sport'] == 'Volleyball') & (df5['City'] == 'Los Angeles')].groupby('Gender').count()['Athlete']

Gender
Men      36
Women    36
Name: Athlete, dtype: int64

In [706]:
# Who won the gold in the 75 - 81KG (Light-Heavyweight) event of Boxing in the Athens Olympics?



In [792]:
df5[(df5['Event'] == '75 - 81KG (Light-Heavyweight)') & (df5['Medal'] == 'Gold') & (df5['City'] == 'Athens')]['Athlete']

25918    WARD, Andre
Name: Athlete, dtype: object